### ISSUE TITLE
Where could I find the example for key generator code?

### ISSUE THREAD

**Issue by wowojyc** [apache/incubator-shardingsphere#2489]
I use raw java code,

`
 KeyGeneratorConfiguration keyGeneratorConfiguration = new 
 KeyGeneratorConfiguration("SNOWFLAKE", "goods_id", new Properties());
 tableRuleConfiguration.setKeyGeneratorConfig(keyGeneratorConfiguration);
`

and I use jpa to insert data in the for loop;
`

        for (int i = 1; i <= 40; i++) {
            Goods goods = new Goods();
            goods.setGoodsName("shangpin" + i);
            goods.setGoodsType((long) (i + 1));
            long goodsId = goodsRepository.save(goods).getGoodsId();
            System.out.println("goods id:" + goodsId);
        }
`

and My database is db_01 and db_02, the table name is goods_0 and goods_1.
The entity code is 
`
@Entity
@Table(name = "goods")
public class Goods {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "goods_id")
    private Long goodsId;

    private String goodsName;
`
But when i call the insert method, I catch the exception 
**A different object with the same identifier value was already associated with the session**

Do I miss some important configuration?

**Comment by wowojyc** [comment]
Addition some extra message
I use Spring JPA under SpringBoot 2.0+, and the sharding jdbc version is 4.0.0-RC1

**Comment by tristaZero** [comment]
Please refer to the configurations in [JPA-EXAMPLE](https://github.com/tristaZero/sharding-sphere-example/tree/dev/sharding-jdbc-example/sharding-example)
@wowojyc

**Comment by wowojyc** [comment]
Thank you for your reply
I have read the example code and In fact the code runs well.

However my scene has a bit of difference.

In my application, I need dynamic shard rule, so I must use raw java code instead of spring boot application.properties. 

I have several questions.
1. If I want use distributed primary key generated by sharding jdbc, how the table column like in database schema? Is it necessary set 'primary key ' or Auto increment attributes?
2. How the JPA entity code like ? 
    @Id or @GeneratorValue is also necessary? If the answer is yes, which strategy is ok

**Comment by tristaZero** [comment]
> If I want use distributed primary key generated by sharding jdbc, how the table column like in database schema? Is it necessary set 'primary key ' or Auto increment attributes?
If you want to use use distributed primary key generated by sharding jdbc, setting 'primary key ' or Auto increment  is not necessary.

>How the JPA entity code like ?
@id or @GeneratorValue is also necessary?
f you want to use use distributed primary key generated by sharding jdbc,  @id or @GeneratorValue is also not necessary.

**Comment by wowojyc** [comment]
The conclusion is that id does not matter if I set primary key with auto increment property and use @id and @GeneratorValue annotations. Am I right?

**Comment by wowojyc** [comment]
I use hibernate annotation @Entity, it requires me to specify identifier and I must add it to one property.

**Comment by wowojyc** [comment]
Add some extra message.
The application runs well if I remove below setting in the application.properteis

**spring.jpa.properties.hibernate.temp.use_jdbc_metadata_defaults=false**

I add this setting because of some hibernate exceptions in the log, but I don't know why this will affect the distribute primary key.

The springboot version is 2.0+

**Comment by tristaZero** [comment]
@wowojyc  Is your program ok now?
When you use the key generator of ShardingSphere, please do not make Springboot generate key. By set the strategy for @GeneratedValue(strategy = GenerationType.IDENTITY) is ok, if @Id is necessary.

If it's possible, can your give the detail of error logs?

**Comment by wowojyc** [comment]
At first.
I meet the exception like this.
**spring jpa - a different object with the same identifier value was already associated with the session**

And I found the hibernate execute "select last_insert_id()' after my normal insert execute successfully. 
After some try, I check my application.properties again.

According to the [解决hibernate执行select last_insert_id问题](https://www.tag.gg/showinfo-14-122887-0.html), I removed the setting and it does work.

The reason why I add this setting is that my database is postgresql and its jdbc connection does not implement createBlob interface.
